### PR TITLE
Removing Articles::Builder making policy decision

### LINKED ISSUE #16529: Address ArticlesController#new's conditional authorize
At present (current main branch as of <2022-02-10 Thu>), we have the following code:

https://github.com/forem/forem/blob/87e1622e6b85132b948367f16262932ff912bc9b/app/controllers/articles_controller.rb#L47-L52

That code allows for something outside the Policy ecosystem to say whether or not to perform an authorize check.  We have the Articles::Builder class that determines whether or not to perform validations:

https://github.com/forem/forem/blob/87e1622e6b85132b948367f16262932ff912bc9b/app/services/articles/builder.rb#L19-L28

The "why" of that code should move into a policy object.

This informs #16483

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Documentation Update

## Description

Removing Articles::Builder making policy decision

This change is a refactoring through triangulation.  Given that
`ArticlePolicy#new?` returned true, I'm prepared to assume that calling
`authorize(Article)` in all cases is acceptable.

So to narrow the Builder making a policy decision I renamed the returned
value to reflect what it was actually doing in the logic.  And in
renaming, flipped the polarity of the boolean.  Why the flip?  Because
`needs_authorization == !store_location`.

In consultation with Allison and Jennie, I'm proceeding with a short-cut
to get me unstuck.  That unstuck is namely "I need to ensure that the
articles#new action can go through authorization."

Given that I'll be spending time in the authorization layer, I hope
these noted short-cuts and comments will be useful in future spelunking
efforts regarding authorization.

## Related Tickets & Documents

Closes #16529

## QA Instructions, Screenshots, Recordings

None, as this should be a no-impact change.

### UI accessibility concerns?

None.

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] I will share this change internally with the appropriate teams

### CODE REVIEW

File: app/policies/article_policy.rb [hunk 1]
```diff
@@ -12,19 +12,47 @@ def self.limit_post_creation_to_admins?
     FeatureFlag.enabled?(:limit_post_creation_to_admins)
   end
 
+  # @note [@jeremyf] I am re-implemnenting the initialize method, but removing the Pundit
+  #       authorization.  There's an assumption that all policy questions will require a user,
+  #       unless you know specifically that they don't.
+  #
+  # @todo [@jeremyf] I don't like altering the initializer and it's core assumption.  But the other
```
Comment (citizen428) [inline app/policies/article_policy.rb:19]:
```suggestion
  # @todo [@jeremyf] I don't like altering the initializer and its core assumption.  But the other
```

**Review by citizen428 — APPROVED** [review]

File: app/policies/article_policy.rb (hunk 1 above)
Comment (jeremyf) [reply app/policies/article_policy.rb:19]: Someday, I'll get that "it's/its" correct.

File: app/policies/article_policy.rb [hunk 2]
```diff
@@ -52,11 +81,16 @@ def permitted_attributes
...
 
   def subscriptions?
+    require_user!
     user_author? || user_super_admin?
   end
 
   private
 
+  def require_user!
+    raise Pundit::NotAuthorizedError, I18n.t("policies.application_policy.you_must_be_logged_in") unless user
```
Comment (dscottS3) [inline app/policies/article_policy.rb:91]: this is just a personal preference but I like the explicit return early
```suggestion
    return if user
    raise Pundit::NotAuthorizedError, I18n.t("policies.application_policy.you_must_be_logged_in")
```

**Review by dscottS3 — APPROVED** [review]

File: app/policies/article_policy.rb (hunk 2 above)
Comment (jeremyf) [reply app/policies/article_policy.rb]: Definitely like guard conditions, and agree that this is a bit nicer

**Pull request merged**